Expose Stream Recording Apis in python

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96775
* #89146
* __->__ #96384


Differential Revision: [D43999891](https://our.internmc.facebook.com/intern/diff/D43999891)

eellison committed:
Expose API to allocate to private pool without doing cuda graph recording

[ghstack-poisoned]

eellison committed:
Update on "Expose API to allocate to private pool without doing cuda graph recording"




[ghstack-poisoned]

eellison commented on test/test_cuda.py:
```diff
@@ -5429,6 +5429,78 @@ def test_resnet(self):
+            all cudagraph tensors in use should be reflected in the allocator or they will be overwritten.
+            existing_graph should already have been used in a capture, and the mem_pool must already exist.
+            """
+            capture_id = existing_graph.id()
+            torch._C._cuda_allocateThreadToPrivatePool(device, mem_pool)
+            torch._C._cuda_notifyCaptureBegin(device, capture_id, mem_pool)
+            try:
+                yield
+            finally:
+                torch._C._cuda_notifyCaptureAboutToEnd(device, capture_id)
```
We need a capture id unique identifier that wont conflict with any other ongoing unique identifiers. The way to generate a unique identifier is through starting a cudagraph stream recording in cuda. 

Here I am just reusing a finished recording id with the graph, we could alternately start and end a cudastream recording, no strong opinions here..

eellison committed:
Update on "Expose API to allocate to private pool without doing cuda graph recording"


Exposes a way for users to do allocations to a private pool without doing a cudagraph capture. We assume that there is already an existing graph that was used to get a unique capture_id identifier, and then use that to reuse the existing apis. We mimic the existing cudagraph capture api by only doing allocations to a particular thread, just as the stream was thread-local.

This will be needed for the subsequent PR which will do warmup in a cudagraph pool. That pr also also adds some runtime checking which helped find a leaked cublas workspace allocation.

[ghstack-poisoned]

eellison committed:
Update on "Expose API to allocate to private pool without doing cuda graph recording"


Exposes a way for users to do allocations to a private pool without doing a cudagraph capture. We assume that there is already an existing graph that was used to get a unique capture_id identifier, and then use that to reuse the existing apis. We mimic the existing cudagraph capture api by only doing allocations to a particular thread, just as the stream was thread-local.

This will be needed for the subsequent PR which will do warmup in a cudagraph pool. That pr also also adds some runtime checking which helped find a leaked cublas workspace allocation.

[ghstack-poisoned]

eellison committed:
Update on "Expose API to allocate to private pool without doing cuda graph recording"


Exposes a way for users to do allocations to a private pool without doing a cudagraph capture. We assume that there is already an existing graph that was used to get a unique capture_id identifier, and then use that to reuse the existing apis. We mimic the existing cudagraph capture api by only doing allocations to a particular thread, just as the stream was thread-local.

This will be needed for the subsequent PR which will do warmup in a cudagraph pool. That pr also also adds some runtime checking which helped find a leaked cublas workspace allocation.

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43999891).

eellison committed:
Update on "Expose API to allocate to private pool without doing cuda graph recording"


Exposes a way for users to do allocations to a private pool without doing a cudagraph capture. We assume that there is already an existing graph that was used to get a unique capture_id identifier, and then use that to reuse the existing apis. We mimic the existing cudagraph capture api by only doing allocations to a particular thread, just as the stream was thread-local.

This will be needed for the subsequent PR which will do warmup in a cudagraph pool. That pr also also adds some runtime checking which helped find a leaked cublas workspace allocation.

Differential Revision: [D43999891](https://our.internmc.facebook.com/intern/diff/D43999891)

[ghstack-poisoned]

eellison committed:
Update on "Expose API to allocate to private pool without doing cuda graph recording"


Exposes a way for users to do allocations to a private pool without doing a cudagraph capture. We assume that there is already an existing graph that was used to get a unique capture_id identifier, and then use that to reuse the existing apis. We mimic the existing cudagraph capture api by only doing allocations to a particular thread, just as the stream was thread-local.

This will be needed for the subsequent PR which will do warmup in a cudagraph pool. That pr also also adds some runtime checking which helped find a leaked cublas workspace allocation.

Differential Revision: [D43999891](https://our.internmc.facebook.com/intern/diff/D43999891)

[ghstack-poisoned]

eellison committed:
Update on "Expose API to allocate to private pool without doing cuda graph recording"


Exposes a way for users to do allocations to a private pool without doing a cudagraph capture. We assume that there is already an existing graph that was used to get a unique capture_id identifier, and then use that to reuse the existing apis. We mimic the existing cudagraph capture api by only doing allocations to a particular thread, just as the stream was thread-local.

This will be needed for the subsequent PR which will do warmup in a cudagraph pool. That pr also also adds some runtime checking which helped find a leaked cublas workspace allocation.

Differential Revision: [D43999891](https://our.internmc.facebook.com/intern/diff/D43999891)

[ghstack-poisoned]

eellison committed:
Update on "Expose API to allocate to private pool without doing cuda graph recording"


Exposes a way for users to do allocations to a private pool without doing a cudagraph capture. We assume that there is already an existing graph that was used to get a unique capture_id identifier, and then use that to reuse the existing apis. We mimic the existing cudagraph capture api by only doing allocations to a particular thread, just as the stream was thread-local.

This will be needed for the subsequent PR which will do warmup in a cudagraph pool. That pr also also adds some runtime checking which helped find a leaked cublas workspace allocation.

Differential Revision: [D43999891](https://our.internmc.facebook.com/intern/diff/D43999891)

[ghstack-poisoned]

eellison committed:
Update on "Expose API to allocate to private pool without doing cuda graph recording"


Exposes a way for users to do allocations to a private pool without doing a cudagraph capture. We assume that there is already an existing graph that was used to get a unique capture_id identifier, and then use that to reuse the existing apis. We mimic the existing cudagraph capture api by only doing allocations to a particular thread, just as the stream was thread-local.

This will be needed for the subsequent PR which will do warmup in a cudagraph pool. That pr also also adds some runtime checking which helped find a leaked cublas workspace allocation.

Differential Revision: [D43999891](https://our.internmc.facebook.com/intern/diff/D43999891)

[ghstack-poisoned]

eellison committed:
Update on "Expose API to allocate to private pool without doing cuda graph recording"


Exposes a way for users to do allocations to a private pool without doing a cudagraph capture. We assume that there is already an existing graph that was used to get a unique capture_id identifier, and then use that to reuse the existing apis. We mimic the existing cudagraph capture api by only doing allocations to a particular thread, just as the stream was thread-local.

This will be needed for the subsequent PR which will do warmup in a cudagraph pool. That pr also also adds some runtime checking which helped find a leaked cublas workspace allocation.

Differential Revision: [D43999891](https://our.internmc.facebook.com/intern/diff/D43999891)

[ghstack-poisoned]

eellison committed:
Update on "Expose API to allocate to private pool without doing cuda graph recording"


Exposes a way for users to do allocations to a private pool without doing a cudagraph capture. We assume that there is already an existing graph that was used to get a unique capture_id identifier, and then use that to reuse the existing apis. We mimic the existing cudagraph capture api by only doing allocations to a particular thread, just as the stream was thread-local.

This will be needed for the subsequent PR which will do warmup in a cudagraph pool. That pr also also adds some runtime checking which helped find a leaked cublas workspace allocation.

Differential Revision: [D43999891](https://our.internmc.facebook.com/intern/diff/D43999891)

[ghstack-poisoned]

eellison committed:
Update on "Expose API to allocate to private pool without doing cuda graph recording"


Exposes a way for users to do allocations to a private pool without doing a cudagraph capture. We assume that there is already an existing graph that was used to get a unique capture_id identifier, and then use that to reuse the existing apis. We mimic the existing cudagraph capture api by only doing allocations to a particular thread, just as the stream was thread-local.

This will be needed for the subsequent PR which will do warmup in a cudagraph pool. That pr also also adds some runtime checking which helped find a leaked cublas workspace allocation.

Differential Revision: [D43999891](https://our.internmc.facebook.com/intern/diff/D43999891)

[ghstack-poisoned]

zdevito commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -1203,6 +1203,25 @@ static void registerCudaPluggableAllocator(PyObject* module) {
 
+  m.def(
+      "_cuda_beginAllocateCurrentStreamToPool",
+      [](int device, at::cuda::MempoolId_t mempool_id) {
```
nit: feels weird to use the default stream but explicit device vs (default device + default stream) or (explicit stream)

eellison replied:
True. Is there any way of passing a python object as `CudaStream_t` ?

zdevito commented on test/test_cuda.py:
```diff
@@ -5511,6 +5511,81 @@ def test_resnet(self):
+                b = int8_cuda(LARGE_BUFFER)
+                del a, b
+
+        graph_thread = threading.Thread(target=use_pool)
```
what is the intention of the threads here? In all likelihood without additional synchronization one is just going to run entirely before the other.  Barriers placed inside the threads would guarantee some interleaving.

zdevito reviewed (approved):
Looks good. Couple minor comments.

eellison committed:
Update on "Expose Stream Recording Apis in python"



Differential Revision: [D43999891](https://our.internmc.facebook.com/intern/diff/D43999891)

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
